lib/honeydew/job_runner.ex

```diff
@@ -0,0 +1,87 @@
+defmodule Honeydew.JobRunner do
+  use GenServer
+  require Logger
+  require Honeydew
+
+  alias Honeydew.Job
+  alias Honeydew.Crash
+  alias Honeydew.Worker
+
+  @doc false
+  def start_link(opts) do
+    GenServer.start_link(__MODULE__, opts)
+  end
+
+  defmodule State do
+    @moduledoc false
+
+    defstruct [:worker, :job, :module, :worker_private]
+  end
+
+
+  #
+  # Internal API
+  #
+
+  def run_link(job, module, worker_private) do
+    GenServer.start_link(__MODULE__, [self(), job, module, worker_private])
+  end
+
+  @impl true
+  def init([worker, %Job{job_monitor: job_monitor} = job, module, worker_private]) do
+    Process.put(:job_monitor, job_monitor)
+
+    {:ok, %State{job: job,
+                 module: module,
+                 worker: worker,
+                 worker_private: worker_private}, {:continue, :run}}
+  end
+
+  defp do_run(%State{job: %Job{task: task} = job,
+                     module: module,
+                     worker: worker,
+                     worker_private: worker_private} = state) do
+    private_args =
+      case worker_private do
+        {:state, s} -> [s]
+        :no_state   -> []
+      end
+
+    result =
+      try do
+        result =
+          case task do
+            f when is_function(f) -> apply(f, private_args)
+            f when is_atom(f)     -> apply(module, f, private_args)
+            {f, a}                -> apply(module, f, a ++ private_args)
+          end
+        {:ok, result}
+      rescue e ->
+        {:error, Crash.new(:exception, e, System.stacktrace())}
+      catch
+        :exit, reason ->
+          # catch exit signals and shut down in an orderly manner
+          {:error, Crash.new(:exit, reason)}
+        e ->
+          {:error, Crash.new(:throw, e, System.stacktrace())}
+      end
+
+    :ok = Worker.job_finished(worker, %{job | result: result})
+
+    state
+  end
+
+
+  @impl true
+  def handle_continue(:run, state) do
+    {:stop, :normal, do_run(state)}
+  end
```

Comment:
I know this is succinct, but it took me a minute to figure out what was going on here. How about writing it like

```
  def handle_continue(:run, state) do
    state = do_run(state)

    {:stop, :normal, state}
  end
```

At first I thought why is it immediately stopping in handle continue, but then after looking at it for a minute I figured it out.

Also, I think running the job in`handle_continue` makes it so you can't receive any other messages. Does that make cancellation or inspecting the running process difficult to do?

Reply:
The first thing I think is you could run the user code in monitored task, but that's kind of the purpose of this module anyways. Not blocking would be nice in the case where you wanted to interrogate a job to get its progress or allow the user to handle a requested job cancellation. Probably something to tackle at another time, though.

Reply:
i mean, the fact is that we have to hand off control to foreign code, and without preemption and context switching, there's no way to get control back.

for job cancellation, the Worker can just hard kill the runner. progress is handled in a cooperative way with the use via https://github.com/koudelka/honeydew/blob/master/lib/honeydew/progress.ex